File: Data Structures & Algorithms/depth-of-binary-tree/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        stack = [(root, 1)]
        max_depth = 0
        while stack:
            node, depth = stack.pop()
            if node:
                stack.extend([(node.right, depth + 1), (node.left, depth + 1)])
                max_depth = max(max_depth, depth)

        return max_depth

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-6.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int dfs(TreeNode* node, int& max_path_sum) {
        if (node == nullptr) return 0;

        int max_left = dfs(node->left, max_path_sum);
        int max_right = dfs(node->right, max_path_sum);

        max_path_sum = max(max_path_sum, max_left + max_right + node->val);

        return max(0, node->val + max(max_left, max_right));
    }

public:
    int maxPathSum(TreeNode* root) {
        if (root == nullptr) return 0;

        int max_path_sum = root->val;
        dfs(root, max_path_sum);
        return max_path_sum;
    }
};
